File: java/bridge/Device.java
/**
 * 实现部分接口（Implementor）：设备。
 * 定义设备最基础的操作，具体由 Tv / Radio 等实现类完成。
 * 遥控器（Abstraction）只依赖这个接口，不关心具体是哪种设备。
 */
public interface Device {
    boolean isOn();

    void turnOn();

    void turnOff();

    int getVolume();

    void setVolume(int percent);

    String getName();
}

File: cpp/bridge/device.cpp
#include "device.h"
#include <algorithm>
#include <iostream>

void TV::power_on() {
    on_ = true;
    std::cout << "  [电视机] 开机，画面点亮" << std::endl;
}
void TV::power_off() {
    on_ = false;
    std::cout << "  [电视机] 关机" << std::endl;
}
void TV::set_volume(int percent) {
    volume_ = std::clamp(percent, 0, 100);
    std::cout << "  [电视机] 音量调整为 " << volume_ << std::endl;
}

void Radio::power_on() {
    on_ = true;
    std::cout << "  [收音机] 开机，开始接收信号" << std::endl;
}
void Radio::power_off() {
    on_ = false;
    std::cout << "  [收音机] 关机" << std::endl;
}
void Radio::set_volume(int percent) {
    volume_ = std::clamp(percent, 0, 100);
    std::cout << "  [收音机] 音量调整为 " << volume_ << std::endl;
}
